docs/design/features/byreflike-generics.md

```diff
@@ -7,22 +7,25 @@ Using ByRefLike types in Generic parameters is possible by building upon support
 
 ## Runtime impact
 
-Supporting ByRefLike type as Generic parameters will impact the following IL instructions:
+Supporting ByRefLike type as Generic parameters will impact the following IL instructions.
 
-- `box` &ndash; Types with ByRefLike parameters used in fields cannot be boxed.
+When used with a ByRefLike type these instructions result in undefined behavior:
+- `box` &ndash; When applied to ByRefLike types used in fields cannot be boxed.
+- `constrained.callvirt` &ndash; If this IL sequence resolves to a method implemented on `object` or a default interface method.
```

Comment:
@jaredpar Are we happy that you can produce programs with undefined behavior with a safe vanilla C#? Is there a prior art like that?

In the current implementation, this will typically produce `InvalidProgramException`. The docs for `InvalidProgramException` say "Generally this indicates a bug in the compiler that generated the program.", so people are likely going to filling bugs on this in the Roslyn repo.

Reply:
> Are we happy that you can produce programs with undefined behavior with a safe vanilla C#? Is there a prior art like that?

Regarding a prior art. I am not sure about the specific exception, but one could run into an exception trying to invoke a default interface implementation which is ambiguous. Compiler guards against this situation when it compiles a type implementing the interface, but things can change after the type is compiled.

Regarding this specific situation. Compiler disallows calling object methods on ByRefLike type parameters. Compiler disallows calling non-virtual interface instance methods on ByRefLike type parameters. Compiler also requires `ref` structs to implement all virtual members of an implemented interface. So, if all parts are compiled in C# and interface members aren't added after the fact, one shouldn't run into a situation when the exception is thrown.

Reply:
> one could run into an exception trying to invoke a default interface implementation which is ambiguous

The exception in this case is not InvalidProgramException and the behavior is fully deterministic. The exception is always thrown by the call itself, and it won't cause the method to fail to execute in its entirely. It is not that different from other dynamic type system violations, like InvalidCastException from castclass that is also fully determistic.

> interface members aren't added after the fact

Adding interface members after the fact has been the main motivation for default interface methods.

I am starting to think we need to make these behaviors fully deterministic in this case too.

Reply:
I agree that the runtime behavior should be deterministic.

On the compiler side, I don't think there's any way the compiler could prevent the situation from happening. Default interface methods can appear after compilation. So this has to be a runtime check.

I would expect that invoking a default interface implementation on ref struct would deterministically produce some well-defined exception, ideally one that precisely explains the situation.

Reply:
> On the compiler side, I don't think there's any way the compiler could prevent the situation from happening.

Agree the compiler can't fully prevent the DIM situation. At the same time though for this to happen the user must within the same method declare: 

1. I want to allow `T` to be a `ref struct`
2. I want to call a DIM method on `T` 

That is not wrong but it is to some degree contradictory / inviting this problem to happen. It's enough of sketchy case that part of the design suggested we could even [warn in this case](https://github.com/dotnet/csharplang/blob/main/proposals/ref-struct-interfaces.md#warn-on-dim-invocation) if we desired (LDM did not decide to do this). 

As long as the runtime has deterministic behavior here, preferably with a message or type that is easy to search on, then I think we are okay because this is a pretty narrow case. It can be added to the existing documentation around DIM.

Reply:
> I am starting to think we need to make these behaviors fully deterministic in this case too.
> I agree that the runtime behavior should be deterministic.

I would like to understand more about these phrases. There are a lot words and phrases being used and I'm really not following all of them. What does deterministic mean in this case? We have multiple modes of execution. Does it fail at JIT/compilation? Does it partially execute? Does it fail during interpretation? It seems given other comments that "undefined behavior" is defined as "may fail to execute partially or in its entirety", is this the definition for all uses of "undefined behavior"?

Reply:
> What does deterministic mean in this case? 

It means that the following test passes, on all runtime variants (JIT, interpretter, AOT, CoreCLR, Mono, ...):

```csharp
Assert.Equal(1, M<S>(false));

Assert.Throws<TBDException>(() => M<S>(true));

int M<T>(T value, bool callInterfaceMethod) where T: allows ref struct
    => callInterfaceMethod ? value.M() : 0;

interface IMy
{
     int M() => 1;
}

ref struct S : IMy
{
}
```

`TBDException` can be either `NotSupportedException` or we can introduced a new dedicated exception for this situation (I am not sure whether it is worth it).

Reply:
Yup that is one possibility. I used the term “deterministic” but I really mean fully defined behavior. I will leave the exact semantics up to you as I don’t have a strong preference. However, my suggestion would be we narrowly scope the exception to the problematic location: the call instruction. 

Specifically, imagine the following was present at compile time:

```C#
M(new S());

void M<T>(T value) where T: I, allows ref struct
{
    value.M1();
    value.M2();
}

ref struct S : I
{
    void M1() {}
    void M2() {}
}
interface I
{
   void M1();
   void M2();
}
```

This is all valid C#. Now imagine that I is in another assembly and adds M3 with a default interface method. I think this program should continue working without throwing an exception. That means, I think the exception should be narrowly scoped to the invocation of a default interface method, not at type load. 

Similarly, imagine the very complex scenario where the top-level M, I, and S are all in different assemblies. I adds an M3 with a default implementation. The M method calls M3 after M2, but S is never recompiled so it never implements M3. I think the defined behavior should be that calls M1 and M2 succeed. It’s only at the call site to M3 where an exception is thrown. 

The reason I like this computational model is that I think it conceptually matches the problem at the user level: the problem with ref structs and default interface implementations is that the receiver requires boxing. Therefore I think the failure should occur at the place where the problem is: boxing the receiver. At the C# level this is nominally the call site, but for IL this could be at the location where we would otherwise box the receiver.

Reply:
> ```csharp
> int M<T>(T value, bool callInterfaceMethod) where T: allows ref struct
>     => callInterfaceMethod ? value.M() : 0;
> ```

> That means, I think the exception should be narrowly scoped to the invocation of a default interface method, not at type load.

I think this is a bad argument and isn't something we should support. I think we're trying to impose runtime semantic behavior that shouldn't exist based on all the previous talk about not wanting to create invalid IL sequences and also about introducing a branching invariant that isn't related to the underlying issue - the DIM.

Based on the current ECMA-355, the above is invalid IL and hence `InvalidProgramException` is appropriate. We decided that we shouldn't change anything which is where this is coming from.

The second is the argument about the branch condition. I really dislike this because it is unlikely to represent the troublesome concept - the DIM. One can imagine a scenario where, using reflection, people handle this case, but the `callInterfaceMethod` predicate will never represent that issue. This is especially true in the example of about someone introducing a new DIM on a type.

I really don't see the practical utility here. With the compiler helpers proposal below, the above IL sequences would never to exposed for ByRefLike types. This means we can leave the description as before and it remains invalid IL when the type is ByRefLike and must be boxed.

Reply:
> With the compiler helpers proposal below, the above IL sequences would never to exposed for ByRefLike types.

I do not think this is correct. The above IL sequence can be exposed to ByRefLike types through libraries evolving independently, and one of the libraries adding default interface method implementation.

`AmbiguousMatchException` is existing prior art for similar situation. The ambiguous match is also something that can be introduced by libraries evolving independently. We could have said that the ambiguous match is invalid IL and leave the behavior undefined. We have not done that. Instead, we have made the behavior well defined, and introduced a dedicated exception for it that is thrown at the callsite with the problem. A program with ambiguous match is valid IL.

Reply:
> The above IL sequence can be exposed to ByRefLike types through libraries evolving independently, and one of the libraries adding default interface method implementation.

It isn't just adding a default inteface method implementation (DIM impl), it also requires adding a call to it. If there was some way that simply adding a DIM impl caused this to occur, I do get it, but disagree it is all that interesting to change behavior.

> We could have said that the ambiguous match is invalid IL and leave the behavior undefined. We have not done that.

That choice though would be adding to ECMA when something became invalid IL. In this case the definition of invalid IL remains the same, we are relaxing it and thus need to add to ECMA. That is the part I pushing back on. The definition as-is covers this case and we should accept that.

I do accept that if independently evolving libraries can introduce this issue with one not having to recompile then I may need to reconsider, but I don't see how that is presently a common enough scenario to be concerned with. There are three (3) steps that need to happen.

1) The `allows ref struct` constraint needs to be added to a type constrained Generic parameter.
2) The use of the type constrained Generic parameter from (1) needs to call a potential DIM impl on a ByRefLike type.
3) The ByRefLike type in (2) needs to implement an interface where a new DIM impl is added.

I agree that it is conceivable that the interface from (3) is updated, (1) is updated, and then issues occur. This seems very clear the causality though and there is nothing that a specific exception is going to clarify in this case. In my opinion, enabling the branch support here actually hides this case far more since the underlying issue DIM on ByRefLike type, the underlying issue, is being blocked by some unrelated type predicate.

I simply don't see how this makes anything safer and personally see this as creating an architectural time bomb for someone to hit later.

Reply:
For context, the current message for `constrained.callvirt` when a DIM or `object` method is called is:

CoreCLR:
> `System.InvalidProgramException: Cannot create boxed ByRef-like values.`

native AOT:
> `System.InvalidProgramException: Common Language Runtime detected an invalid program. The body of method '<METHOD NAME>' is invalid.`

Reply:
> I agree that it is conceivable that the interface from (3) is updated, (1) is updated, and then issues occur.

Right. It means that user can end up in undefined behavior by doing a sequence of compatible changes. If a sequence of compatible changes leads to an undefined behavior in a vanilla C#, it is a flaw in the design of the feature.

> an architectural time bomb for someone to hit later.

I do not see how this is an architectural timebomb. We just need to declare this behavior to well-defined in the spec and update the runtimes to behave according to the spec. Do you see that as particular hard to do?

Reply:
> If a sequence of compatible changes leads to an undefined behavior in a vanilla C#, it is a flaw in the design of the feature.

My statement "issues occur" is about the unexpected throwing of `InvalidProgramException`. I think we're talking past one another and I accept this is because of what I've written about "undefined behavior". There is nothing undefined here. An `InvalidProgramException` is going to be thrown because the IL is invalid in this scenario - it is fully defined. This means we are remain safe here.

If we have agreement with the above, then (a) I need to remove the "undefined behavior" in the design document and (b) we can then have a discussion about whether we want a more specific exception or not.

For (a), I am going to be adding additional tests that allows us to confirm the defined behavior.

Reply:
I would like to see the exact timing to be defined too. I think that the most sensible behavior is to throw at the callsite.

I do not think that it should throw InvalidProgramException. InvalidProgramException suggests a bug in the compiler (or a bug in unsafe code). It is not the case here.

Reply:
> I would like to see the exact timing to be defined too.

I disagree, but it doesn't look like we are going to align on this. If I were to adopt this perspective, throwing at the callsite is reasonable.

> I do not think that it should throw InvalidProgramException. InvalidProgramException suggests a bug in the compiler (or a bug in unsafe code). It is not the case here.

Again, I disagree, but I'm not sure how to bridge this gap at present. I'll update this design to insert a placeholder.

Reply:
Ok. Do you agree or disagree with the following points?

1. A sequence of compatible changes (compatible from the type system point of view) should not lead to an invalid program or undefined behavior in a pure C#.  
2. Ambiguous default interface method resolution is very similar situation as default interface method invocation on byref-like type.
3. Throwing at the callsite to handle ambiguous default interface method resolution has been the right design. (If we can do it again, we would do it the same way.)

These points are behind my though process.

Reply:
**TL;DR** I see your point now. I agree with your preference here on both the documentation of time of failure and call-site behavior.

I agree with all those points. I was struggling with (1) for a bit, but that was a confusion on my part with the proposed scenario - playing SharpLap more helped. I asked around and now I get the issue mentioned in (2), it makes me sad. Point (3) is where the main mental hurdle remained; however, given my new understanding of (2), point (3) follows directly. I will update the design as suggested, appreciate the patience.